### PR TITLE
Remove cache

### PR DESCRIPTION
**Pull request by husanjun**
The cache files generated by the old code should be deleted after the upgrade

### CODE REVIEW

**Comment by jfcherng** [comment]
I feel what `on_settings_read()` and `install_or_update()` do are duplicate. Do we really need both of them?

**Comment by husanjun** [comment]
> I feel what `on_settings_read()` and `install_or_update()` do are duplicate. Do we really need both of them?

Yes, When the `install_or_update()` method is executed, the configuration is not reread, so update configuration invalid in `install_or_update()`

**Comment by jfcherng** [comment]
> > I feel what `on_settings_read()` and `install_or_update()` do are duplicate. Do we really need both of them?
> 
> Yes, When the `install_or_update()` method is executed, the configuration is not reread, so update configuration invalid in `install_or_update()` , I've tried

But if you delete the cache dir via shutil during install/update, when the install/update is done, the cache dir doesn't exist. Then things in `on_settings_read()`, which suggest intelephense to delete the cache, shouldn't necessary?

**Comment by husanjun** [comment]
> But if you delete the cache dir via shutil during install/update, when the install/update is done, the cache dir doesn't exist. Then things in `on_settings_read()`, which asks intelephense to delete the cache, shouldn't necessary?
 
I'm not sure, because I looked at this:
https://github.com/bmewburn/vscode-intelephense/blob/bd8bd154dddfb117b62f61bdbd825cd6629b7a35/src/extension.ts#L78-L86
But I suppose you're right,  I'll fix it right away

File: plugin.py [hunk 1]
```diff
@@ -43,6 +44,18 @@ def get_additional_variables(cls) -> Optional[Dict[str, str]]:
...
+    def install_or_update(cls) -> None:
+        super().install_or_update()
+        settings, _ = cls.read_settings()
+        cache_path = settings.get("initializationOptions", {}).get("storagePath", "")
+        variables = sublime.active_window().extract_variables()
+        variables.update(cls.additional_variables())
+
+        cache_path = os.path.realpath(os.path.abspath(sublime.expand_variables(cache_path, variables)))
+        if cache_path and os.path.isdir(cache_path):
+            shutil.rmtree(cache_path, ignore_errors=True)
```
Comment (rchl) [inline plugin.py:57]: You could shortcut the code and do nothing if the initial value of `cache_path` is falsy.

But also there are more fundamental issues with this logic:
 - This doesn't consider project overrides where the path might be different.
 - If the path is unset then I guess intelephense still saves the cache to the default location but then we won't clear it?
 - I think ideally we would just pass `clearCache` when appropriate but I think "appropriate" might be hard to determine currently. I would need to invest more time into this to be able to maybe figure out a better solution.

**Comment by rchl** [comment]
Also this code seems very dangerous. If someone would accidentally set '~' or `/` as the path then I'm guessing bad things would happen.

File: plugin.py (hunk 1 above)
Comment (rchl) [reply plugin.py:57]: Also this code seems very dangerous. If someone would accidentally set '~' or `/` as the path then I'm guessing bad things would happen.

Comment (husanjun) [reply plugin.py:57]: > Also this code seems very dangerous. If someone would accidentally set '~' or `/` as the path then I'm guessing bad things would happen.

We should add Intelephense after the user directory to ensure directory security

Comment (jfcherng) [reply plugin.py:57]: > * I think ideally we would just pass `clearCache` when appropriate but I think "appropriate" might be hard to determine currently. I would need to invest more time into this to be able to maybe figure out a better solution.

^ this +1

Comment (husanjun) [reply plugin.py:57]: > > * I think ideally we would just pass `clearCache` when appropriate but I think "appropriate" might be hard to determine currently. I would need to invest more time into this to be able to maybe figure out a better solution.
> 
> ^ this +1

Okay, but the old cache files will always exist and grow

Comment (jfcherng) [reply plugin.py:57]: Ah, wait, `clearCache` only clean the current project but not all project?

Comment (husanjun) [reply plugin.py:57]: > Ah, wait, `clearCache` only clean the current project but not all project?

Yes, The server will occasionally create a directory similar to the one in the screenshot

image

`clearcache` only clears the files in the directory, it does not delete the directory, so the directory grows

Comment (husanjun) [reply plugin.py:57]: > > Ah, wait, `clearCache` only clean the current project but not all project?
> 
> Yes, The server will occasionally create a directory similar to the one in the screenshot
> 
> image
> 
> `clearcache` only clears(It's actually an update) the files in the directory, it does not delete the directory, so the directory grows

These two directories are the same project, and I guess that's why bmewburn/vscode-intelephense#1523  can't be provided

**Pull request closed without merging**